mysys/my_uuid_v7.c

```diff
@@ -0,0 +1,157 @@
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1335  USA */
+
+/*
+  implements Universal Unique Identifiers version 7, as described in
+  draft-ietf-uuidrev-rfc4122bis-14.
+
+  A UUIDv7 has the following structure:
+
+  Field                     Octet #          Note
+ unix_ts_ms                 0-5     Big-endian unsigned number of
+                                    Unix epoch timestamp in
+                                    milliseconds.
+ version_and_rand_a         6-7     Pseudo-random 12 bits multiplexed
+                                    with the version number.
+ variant_and_rand_b         8-15    Pseudo-random 62 bits multiplexed
+                                    with the version number.
+*/
+
+#include "mysys_priv.h"
+#include <my_rnd.h>
+#include <m_string.h>
+#include <myisampack.h> /* mi_int2store, mi_int4store */
+
+static my_bool my_uuid_v7_inited= 0;
+static struct my_rnd_struct uuid_rand;
+static uint64 milliseq;
+static uint64 uuid_time= 0;
+
+static mysql_mutex_t LOCK_uuid_v7_generator;
+
+#define UUID_VERSION      0x7000
+#define UUID_VARIANT      0x8000000000000000
+
+/**
+  Init structures needed for my_uuid_v7
+
+  @func my_uuid_v7_init()
+  @param seed1		Seed for random generator
+  @param seed2		Seed for random generator
+
+  @note
+    Seed1 & seed2 should NOT depend on clock. This is to be able to
+    generate a random mac address according to UUID specs.
+*/
+void my_uuid_v7_init(uint32 seed1, uint32 seed2)
+{
+  uint64 now;
+
+  if (my_uuid_v7_inited)
+    return;
+  my_uuid_v7_inited= 1;
+  now= my_interval_timer()/100;
+  milliseq= 0;
+
+  my_rnd_init(&uuid_rand, (uint32) (seed1 + now), (uint32) (now/2+ getpid()));
+  mysql_mutex_init(key_LOCK_uuid_v7_generator, &LOCK_uuid_v7_generator, MY_MUTEX_INIT_FAST);
+}
+
+/**
+   Create a global unique identifier version 7(uuidv7)
+
+   @func  my_uuid_v7()
+   @param to   Store uuidv7 here. Must be of size MY_UUID_SIZE (16)
+*/
+void my_uuid_v7(uchar *to)
+{
+  uint64 tv, unix_ts_ms;
+  uint16 rand_a, version_and_rand_a;
+  uint64 rand_b, variant_and_rand_b;
+
+  DBUG_ASSERT(my_uuid_v7_inited);
+
+  mysql_mutex_lock(&LOCK_uuid_v7_generator);
+  tv= my_hrtime().val / 1000;
+
+  if (likely(tv > uuid_time))
+  {
+    /*
+      Current time is ahead of last timestamp, as it should be.
+      If we "borrowed time", give it back, just as long as we
+      stay ahead of the previous timestamp.
+    */
+    if (milliseq)
+    {
+      uint64 delta;
+      DBUG_ASSERT((tv > uuid_time) && (milliseq > 0));
```

Comment:
`millisec > 0` is implied by the `if` condition.

Is the absolutely no possibility of and overflow while this is here? Its looking like a risky assertion to me, but I could be missing something.